src/System.Net.Http/src/System/Net/Http/Headers/HeaderUtilities.cs

```diff
@@ -92,14 +98,12 @@ internal static bool IsInputEncoded5987(string input, out string output)
                     foreach (byte b in bytes)
```

Comment:
This is an expensive way to get the UTF8 bytes, allocating a string for each character and then a byte[] for each string.  I realize the purpose of your PR isn't to change this, but if this code path is actually being used on each request, it'd be great to address this, either now or as part of a separate PR.

@GrabYourPitchforks, with all of the APIs you've added around Rune and UTF8, what's the simplest / cheapest way to do this?

Reply:
The easiest / cheapest way to do this would be to rent a `byte[]` at the beginning of the method, call `Encoding.UTF8.GetBytes(originalString, rentedArray, ...)` to bulk convert the entire thing from UTF-16 to UTF-8, then iterate through the resulting array byte-by-byte and either copy as-is or hex-encode as necessary. The worst-case length of the `byte[]` you need to rent is 3x the size of the input string, but given that the input strings are expected to be small this shouldn't be a huge issue.

(The code as it currently exists is incorrect because it doesn't handle surrogate characters correctly.)

Reply:
Thinking more on my comment here, I think it would be worthwhile for us to do a sweep of Framework code to see if any other components are incorrectly transcoding to UTF-8 char-by-char. Any while nobody has yet reported a transcoding bug in this particular component, it's the third time in the very recent past I've seen the same incorrect pattern being used.

I'd not hold up this PR while trying to address that issue, though. Ideally we'd spend some time auditing all of corefx code and send a single large PR that corrects everything in one shot.